flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.streaming.api.functions.source;
+
+import org.apache.flink.api.common.io.FileInputFormat;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.fs.FileInputSplit;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.JobException;
+import org.apache.flink.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This is the single (non-parallel) task which monitors a user-provided path and assigns splits
+ * to downstream tasks for further reading and processing. Which splits will be further processed
+ * depends on the user-provided {@link FileSplitMonitoringFunction.WatchType}.
+ */
+public class FileSplitMonitoringFunction<OUT>
+	extends RichSourceFunction<FileInputSplit> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileSplitMonitoringFunction.class);
+
+	/**
+	 * Specifies when computation will be triggered.
+	 */
+	public enum WatchType {
+		REPROCESS_WITH_APPENDED		// Reprocesses the whole file when new data is appended.
+	}
+
+	/** The path to monitor. */
+	private final String path;
+
+	/** The default parallelism for the job, as this is going to be the parallelism of the downstream readers. */
+	private final int readerParallelism;
+
+	/** The {@link FileInputFormat} to be read. */
+	private FileInputFormat<OUT> format;
+
+	/** How often to monitor the state of the directory for new data. */
+	private final long interval;
+
+	/** Which new data to process (see {@link WatchType}. */
+	private final WatchType watchType;
+
+	private long globalModificationTime;
+
+	private FilePathFilter pathFilter;
+
+	private volatile boolean isRunning = true;
```

Comment:
In the Kafka source we have it like this as well. I think if it is set in open this can lead to race conditions with cancel being called at weird places and stuff. I vaguely remember that that was the reason for why it is like this in the Kafka source.